### PR TITLE
Keep User Logged In: Export key for encrypted key login

### PR DESCRIPTION
**Pull request by darkwing**
This is an alternative to https://github.com/MetaMask/KeyringController/pull/147 , whereby we create an encryption key based off of the derived key of browser-passworder

Depends on: https://github.com/MetaMask/browser-passworder/pull/20

### CODE REVIEW

File: index.js
Comment (darkwing) [inline index.js:648]: Nit: Make sure we don't need this update if this value didn't change.

File: index.js
Comment (darkwing) [inline index.js:190]: Nit: Do we need this call?

File: index.js
Comment (darkwing) [reply index.js:648]: Confirmed we *do* need this.

File: index.js
Comment (naugtur) [inline index.js:182]: (could) document (with a comment or in jsdoc)  the fact that setLocked can have no asynchronous side-effects. We don't want to end up running unlockKeyrings and those sideeffects in non-deterministic order.

File: package.json
Comment (Gudahtt) [inline package.json:36]: I'd recommend using a commit hash here rather than a branch. I've found in the past that referencing a branch can result in confusing situations where an older commit is still used even after the branch is updated, because Yarn has cached the resolution of this branch to the older commit. Not 100% sure whether it still works this way though.

**Comment by Gudahtt** [comment]
How will this work with mobile? Mobile's decryptor doesn't have any equivalent function to `encryptWithDetail`/`decryptWithDetail`, nor does it have any need for it.

File: package.json
Comment (darkwing) [reply package.json:36]: You're right, we do still get weird cached commit problems.  I stuck with branch as I was making frequent changes and constantly update the commit message was frustrating.

Comment (Gudahtt) [reply package.json:36]: Fair enough!

File: index.js
Comment (legobeat) [inline index.js:556]: Realizing I'm probably late to the party here but: How ingrained is the use of the word "password" in user-facing messages wrt this passphrase?

The reason why I'd like to be meticulous and (however late) make clear distinctions is to help users set expectations:
* A `password` is validated against some source of truth to perform authorization. There is always *some* way to reset it to regain access.
* A `passphrase` is used to derive an encryption key. If it's lost it and there are no backups, there is no recourse.

File: index.js
Comment (legobeat) [inline index.js:575]: An alternative approach (which should not require API changes to `browser-passworder`) would be to generate the key itself here (if it does not exist) and use the existing `encryptWithKeyMethod`. That would reduce the code complexity and reduce the number of potential states.

Was this considered, and if so, why is it not preferred?

File: index.js
Comment (legobeat) [inline index.js:187]: While the passphrase is already set as a private variable (so shouldn't be any new exposure introduced by this PR), this sets of bells for me.

Could we get away without keeping it in memory (since the decrypted key seems to be as well)? If not, could we at least encapsulate it further? (cf. https://github.com/MetaMask/eslint-config/pull/244)

It seems reasonable that even if so, that can be addressed in a separate PR from this.

File: index.js
Comment (Gudahtt) [reply index.js:575]: This package has to work across both mobile and extension, and they use entirely different encryption methods. Mobile uses a native AES API, whereas extension uses the web crypto API. This is managed by keeping all cryptography in the "encryptor" instance, which is injected into this class.

Comment (Gudahtt) [reply index.js:575]: Could you elaborate on how that would reduce the number of potential states? I don't follow that part.

File: index.js
Comment (darkwing) [reply index.js:556]: I'm happy to update to either word, but I don't think it should block this PR.

File: index.js
Comment (Gudahtt) [inline index.js:601]: What is the relationship between these two updates? It's not clear to me why this order is important.

**Comment by Gudahtt** [comment]
The `cacheEncryptionKey` option looks good! Simpler than I thought it would be

File: index.js
Comment (darkwing) [reply index.js:601]: Grrrr.  It's a weird extension issue.  If I call `this.memStore.updateState({ encryptionKey: newExportedKeyString });` *before* `_updateMemStoreKeyrings`, the account is created in the keyring *but* somehow isn't being propagated to the extension before the Preferences controller tries to `syncAddresses`.

Comment (darkwing) [reply index.js:601]: The visual side effect is that the new account wouldn't be selected in the extension -- Account 1 would be selected.  When you open up the accounts menu, however, it's there.

File: index.js
Comment (legobeat) [reply index.js]: I was loosely referring to branching complexity. Inlining the logic in `encryptWithDetail` could reduce the call sites into `encryptor`, parallel logic, and code surface area. Consider:

```
        ___
_______/   \______________
       \___/
```

vs

```
     _____    ________
____/     \__/        \____
    \_____/  \________/
```

Conversely for decrypt.

 It's not a large number of codes we're talking about here, but when juggling passphrases and exposed private keys around it's prudent to be more picky IMO.

Comment (Gudahtt) [reply index.js]: Ah I see. I don't think it would reduce branching complexity overall, but it would shift it from that package into this one.

Either way, we will have to persist the keyring state with a password and no key, or with a key and no password. And the same goes for unlock: sometimes we'll have a password but no key, and sometimes we'll have a key but no password. That's two paths each, for unlock and persist.

That's four paths for the extension at least. Handling mobile as well is more complex, as their encryptor doesn't yet support returning a key (it has no immediate need to support this). So that's an additional path for unlock and persist (where a password is supplied, but we don't want to ask for a key and cache it).

Even if we inline some of the encryptor changes here, we would still have those 6 branches, the same as we have now.

Comment (Gudahtt) [reply index.js]: We could eliminate two of those branches by updating the mobile encryptor to also support `encryptWithDetail`. That'd make it easier for us to combine it with `encrypt` and eliminate the extra two methods on the encryptor as well.

But it's unclear to me how that would benefit mobile. I guess it would let them keep the key in-memory and let the garbage collector cleanup the password, which.... seems preferable but it's hard to argue it has any tangible benefit to security beyond allowing us to simplify some of this code. Postponing that change for later makes sense to me, especially given that a lot of things are waiting on this change (_a lot_ of MV3 bugs surrounding the restart of the service worker are linked to this work).

Comment (Gudahtt) [reply index.js]: If we just wanted to reduce the API surface of the encryptor, that would be a bit easier to accomplish. We could update `encrypt` and `decrypt` to return an object, and add the details to that rather than adding the new `encryptWithDetails`/`decryptWithDetails` functions.

That'd make the extension encryptor act differently from the mobile encryptor for those methods, so we'd need to handle that difference or change the mobile one as well. But that wouldn't be too difficult either way. I expect we'll want to do this eventually.

File: index.js
Comment (naugtur) [inline index.js:144]: Why does this need to be called twice? Looks like the initial call is redundant or there's not enough separation of concerns for addNewKeyring to work regardless.

File: index.js
Comment (naugtur) [inline index.js:557]: What's the difference between holding something in memory store vs class field? Why do we use/need both?

File: index.js
Comment (naugtur) [reply index.js:605]: This comment could be covering the why insteadof/alongwith the what

File: index.js
Comment (naugtur) [inline index.js:659]: (nit) This if structure looks similar across both functions. I'd be tempted to create a function that accepts state and descriptively named functions as reactions to certain states.

File: index.js
Comment (naugtur) [reply index.js:605]: Also, could a test cover this to guarantee it's not undone?

**Review by naugtur — APPROVED** [review]
Left some comments, but none of them are blockers. 
Please consider addressing them before merging, but you have my approval already. This is Fiiiine!

File: package.json
Comment (legobeat) [inline package.json:36]: Before merging this PR, those changes should ideally be merged to `MetaMask/browser-passworder`. If there is a blocker for this, let's make it a branch of the upstream repo and point at the commit hash? 

The reasoning is to not have builds produce different results depending on the current state of branch in an individual's external repo.

File: index.js
Comment (darkwing) [reply index.js:557]: Putting the `encryptionKey` in the memory store bubbles it up to the extension, which allows us to store it in session storage: https://github.com/MetaMask/metamask-extension/pull/15558/files#diff-6fbff2cfe97ac01b77296ef2122c7e0a5b3ff6a84b584b4d1a87482f35eea3d6R3941

Comment (Gudahtt) [reply index.js:557]: We don't generally have a use for _public_ class fields; the password field should be private. That's a change I was going to make in a separate PR at some point; it would be a breaking change because `@metamask/controllers` uses this field today, but they don't need to.

File: package.json
Comment (Gudahtt) [reply package.json:36]: This has just been published to npm as v4.0.1

File: index.js
Comment (Gudahtt) [inline index.js:91]: Hmm. This change caught my eye because it's a functional change (the previous `this.setUnlocked.bind()` wouldn't actually call `this.setUnlocked()`). But it looks like it was supposed to; this was an accidental change introduced in the async/await PR.

File: index.js
Comment (darkwing) [reply index.js:144]: Heh, it's from the first ever commit:
FirstCommit

File: index.js
Comment (Gudahtt) [inline index.js:185]: What motivated this change? We should only be calling `submitPassword` when the keyring is already locked, so this might not be entirely necessary.

I ask because this is introducing a new state update, as a side-effect of `setLocked`, and I'm not sure what effect it will have. And to make matters worse, there is an asynchronous step in `setLocked` which could cause confusing (the `lock` event might end up being emitted after unlocked, which is confusing).

Comment (darkwing) [reply index.js:185]: It's not something I consider absolutely necessary but allowing to submit password, while already unlocked and populated, seems weird.  Like, if you go to a login page but you're already logged in, you get redirected to the home screen.  Like I said, not strictly necessary, but seems something should happen here.

File: index.js
Comment (Gudahtt) [reply index.js]: > If I call `this.memStore.updateState({ encryptionKey: newExportedKeyString });` *before* `_updateMemStoreKeyrings`, the account is created in the keyring *but* somehow isn't being propagated to the extension before the Preferences controller tries to `syncAddresses`.  

Is this also why the calls to `_updateMemStoreKeyrings` were removed in a number of places? I was curious about that as well.

File: index.js
Comment (darkwing) [reply index.js]: I've removed this, as it isn't strictly related to this PR.

File: index.js
Comment (Gudahtt) [reply index.js:545]: I have separated these changes into a new PR to make this one easier to review: https://github.com/MetaMask/KeyringController/pull/154

File: index.js
Comment (Gudahtt) [reply index.js:91]: I have moved this change to a separate PR, and added a test: https://github.com/MetaMask/KeyringController/pull/155

File: index.js
Comment (Gudahtt) [reply index.js]: Huh this is pretty tricky. I think I follow though.

Whenever `memStore` is updated, it'll trigger some other things to run (including `syncAddresses`). By setting the encryption key here before calling `_updateMemStoreKeyrings`, these state update get triggered with an _outdated_ set of keyrings and associated accounts.

An update with stale keyrings/accounts is a problem because it has side-effects like choosing a new selected address in some cases. Whereas a stale (i.e. unset) encryption key is not a problem at all. It just gets set in the next update.

File: index.js
Comment (Gudahtt) [reply index.js]: Another approach to consider would be adding a `try...catch`, and locking if we get an error. That would avoid any misleading `lock` event in the happy path

File: index.js
Comment (Gudahtt) [inline index.js:595]: There is one more instance of `_updateMemStoreKeyrings` that can be removed in `forgetKeyring` (it's right after a call to `persistAllKeyrings`, so it's redundant)

File: index.js
Comment (Gudahtt) [inline index.js:639]: Nit: We can refactor this to only parse the vault once:

```suggestion
        const parsedEncryptedVault = JSON.parse(encryptedVault);
        vault = await this.encryptor.decryptWithKey(
          key,
          parsedEncryptedVault,
        );

        this.encryptionSalt = parsedEncryptedVault.salt;
```

Probably nicer to rename `parsedEncryptedVault` to `encryptedVault`, and rename the current `encryptedVault` to `serializedEncryptedVault` (or something like that), but that was harder to do in a suggestion.

File: index.js
Comment (Gudahtt) [inline index.js:596]: Hmm. Should we keep the `encryptionSalt` in memory alongside the `encryptionKey`? That would help us verify whether the `encryptionKey` matches what was used to encrypt the vault.

Every time a password is used to encrypt the vault, a new encryption key + salt is generated. Once that new encrypted vault is persisted, the previous `encryptionKey` becomes obsolete - it was generated using a different salt, so it will not work to decrypt the vault anymore.

The only scenario where this might happen is if the vault is persisted 3 lines up from here, but then the service worker dies before setting the encryption key in session storage. In that scenario, IF there was already an encryption key set previously, we'd have an invalid key in session storage and the next login attempt would fail.

However if we set the salt in session storage as well, we can compare the session salt to the vault salt to validate the key prior to using it. Then we can bump the user to the login screen in this edge case, and avoid leaving them stuck at the login screen.

File: index.js
Comment (darkwing) [reply index.js:594]: Good shout!

File: test/lib/mock-encryptor.js
Comment (Gudahtt) [inline test/lib/mock-encryptor.js:19]: It'd be nice to have test coverage for a mobile encryptor as well. Maybe we can leave this as-is, and move these modifications to a new mock encryptor module? Or set two up here, and return them as named exports rather than as the default export.

Then we can ensure things blow up properly if `cacheEncryptionKey` is enabled with an encryptor that doesn't support it.

Comment (Gudahtt) [reply test/lib/mock-encryptor.js:19]: Converting this to a class might make that easier, as was done here: https://github.com/MetaMask/controllers/blob/main/tests/mocks/mockEncryptor.ts

Then we can extend the base encryptor to create the one with the browser-passworder additional methods.

File: test/index.js
Comment (Gudahtt) [inline test/index.js:475]: This should probably be checking that the encryption key is set.

File: test/index.js
Comment (Gudahtt) [inline test/index.js:474]: Maybe `cacheEncryptionKey` is a better description of this section? Only a few tests are about the `submitEncryptionKey` function, but they all seem to relate to this option.

File: test/index.js
Comment (Gudahtt) [inline test/index.js:547]: The `cacheEncryptionKey` option isn't being set here, so this assertion isn't doing what it appears to be doing. It'll be `null` here not because it was cleared, but because it was never set.

File: test/index.js
Comment (Gudahtt) [inline test/index.js:500]: We could use an additional test where a _valid_ but _incorrect_ encyption key is passed in. I wouldn't call `{}` an "incorrect" key, it's not a valid key, making this more of a validation test.

File: test/index.js
Comment (Gudahtt) [inline test/index.js:493]: This might be a more robust test if we locked the keyring controller first before submitting the encryption key. Otherwise this keyring state might be from the initial keyring added in the `beforeEach`. It's not clear it's a result of entering the key.

Comment (Gudahtt) [reply test/index.js:493]: That would also let you get rid of the sinon stub, as you wouldn't need to verify the encryptor was called. The state would prove the key was used.

Though... the password is still set. Hmm.  Ideally the keyring controller would be constructed per-test, so you could do setup differently here and never enter a password. But that would be a larger refactor. Maybe in the meantime you can set `keyringController` to `undefined`, and we can refactor the tests separately.

File: test/index.js
Comment (Gudahtt) [inline test/index.js:513]: If you just want to see how a function was called, you don't need to stub it out. Instead use `sinon.spy`, which lets you spy on the function without changing how it behaves.

```suggestion
      const spy = sinon.spy(keyringController, 'persistAllKeyrings');
```

Comment (Gudahtt) [reply test/index.js:513]: Though... generally it's not a good idea to be using internals in tests. They can serve as an unnecessary obstacle to refactoring, and generally they're farther away from testing real use cases. If we could update the assertions to avoid using internals altogether that would be idea.

In this case, the `memStore` can be inspected to see how many accounts there are - there's no need to use a spy. And the same goes for the addition of a new keyring and the removal of an account. It's all in the memStore.

File: index.js
Comment (Gudahtt) [inline index.js:544]: It'd be great to have test coverage for this error

File: index.js
Comment (Gudahtt) [inline index.js:571]: It looks like this branch isn't tested, because all of the tests have `this.password` set.

File: test/index.js
Comment (darkwing) [reply test/index.js:562]: Good shout!  Updated!

File: test/index.js
Comment (Gudahtt) [inline test/index.js:481]: Nit: In these cases where w want to ensure we have a non-empty string, we can use `expect(...).toStrictEqual(expect.stringMatching('.+'))`. This is a bit more specific than the truthy check in that it ensures it's a string, and it avoids the need for this ignore comment.

File: test/index.js
Comment (darkwing) [reply test/index.js]: Good shout!

File: test/index.js
Comment (darkwing) [reply test/index.js:481]: Excellent, I had no idea!

File: index.js
Comment (darkwing) [reply index.js:578]: Fixed!

File: test/index.js
Comment (darkwing) [reply test/index.js]: Done!

File: test/index.js
Comment (Gudahtt) [reply test/index.js:481]: There is one more place you could use this (on this line :sweat_smile: )

File: test/index.js
Comment (Gudahtt) [inline test/index.js:489]: Nit: IMO this would be a better test with this removed. The keyrings-based assertion is enough to show this works; this part just makes the test more fragile by tying it to internals.

File: test/index.js
Comment (Gudahtt) [inline test/index.js:536]: The ESLint error can be avoided by making this the first step

**Review by Gudahtt — DISMISSED** [review]
LGTM! Still a lint failure to resolve, but this all looks great and I can re-approve after that's fixed.

**Review by Gudahtt — APPROVED** [review]

**Pull request merged**
